### ISSUE TITLE
Abort errors in data stream not handled

### ISSUE THREAD

**Issue by Godrules500** [vercel/ai#4101]
### Description

I am on ai-sdk/google-vertex 2.0.8 and ai 4.0.18, but when I call stop() it doesn't hit onFinish, but instead hits onError and returns 
`Invalid state: Controller is already closed`

**Comment by lgrammel** [comment]
Hi can you provide more code for the reproduction?

**Comment by Godrules500** [comment]
Yes, sorry!
  ```
let {
    messages,
    append,
    setMessages,
    isLoading,
    input,
    setInput,
    handleInputChange,
    handleSubmit,
    error,
    reload,
    data,
    setData,
    stop
  }: CustomUseChat = useChat({
    id: chatId,
    initialMessages,
    sendExtraMessageFields: true,
    keepLastMessageOnError: true,
    body: { chatId: chatId, title: chat.title },
    async onResponse(response) {
      if (response.status === 401) {
        // Handle unauthorized
        toast.error(response.statusText)
      }
    },
    onFinish: () => {
      mutate(getChatHistoryPath(user?.id!))
    },
    async onError(error: Error) {
      console.error(error)
      try {
        const errObj = JSON.parse(error.message)
        setMessages(prevMessages => [...prevMessages, errObj])
        await sleep(0)
        console.error(errObj)
      } catch (error) {
        console.warn('Failed to parse JSON:', error)
      }
    },
    maxSteps: 10
  })

```
passed stop through a few components

```
<PromptForm
            user={user}
            chatId={chatId}
            isAtBottom={isAtBottom}
            scrollToBottom={scrollToBottom}
            handleSubmit={handleSubmit}
            handleInputChange={handleInputChange}
            setInput={setInput}
            input={input}
            attachments={attachments}
            setAttachments={setAttachments}
            stop={stop}
          />
          
  const handleInputSubmit = async (e: FormEvent<HTMLFormElement>) => {
    e.preventDefault()
    if (isLoading) {
      stop()
      return
    }
})
```
Then on the route.ts side
     
``` return createDataStreamResponse({
        execute: async streamingData => {
          const result = streamText({
            temperature: 1,
            // experimental_transform: smoothStream(),
            tools,
            model: await setupVertex({ useSearchGrounding: isUseGrounding }),
            system: system_prompt,
            messages: aiMessages,
            maxSteps: 10,
            frequencyPenalty: 0.15,
            onFinish: async ({ response, finishReason, text }) => {
				// Never makes it here.
            }
          })

          result.mergeIntoDataStream(streamingData)
        },
        onError: error => {
		  // Makes it here with the error described above.
          console.error(error)
          const errorObj = handleUnexpectedStreamingErrors(error, chat, true, title)
          chat.messages = [...messages, { ...errorObj }]
          chat.title = title || chat.title
          addOrUpdateChat(chat)
          return JSON.stringify(errorObj)
        }
      })
```

**Comment by lgrammel** [comment]
Hypothesis: abort signals close the client <> next server connection, thus terminating the stream. in the meantime, the server <> llm stream continues and then throws this error.

**Comment by lgrammel** [comment]
## Identify issue type

Identified as **Bug**  
Confidence: **High**

### Rationale
This is a confirmed defect in existing abort behavior: stopping a stream led to an unexpected closed-controller error rather than handling the abort normally.

### Evidence
- Calling useChat.stop() caused onError to receive "Invalid state: Controller is already closed" instead of cleanly handling the abort.
- The server-side LLM stream continued after the client connection was aborted and attempted to write to the closed data stream.
- A maintainer reproduced the issue, identified it as a general data-stream abort-handling error, and stated it was fixed in ai@4.0.20.
